src/compiler/binder.ts

```diff
@@ -264,6 +265,14 @@ namespace ts {
                     let functionType = <JSDocFunctionType>node.parent;
                     let index = indexOf(functionType.parameters, node);
                     return "p" + index;
+                case SyntaxKind.JSDocParameterTag:
+                    Debug.assert(node.parent.kind === SyntaxKind.JSDocCallbackType);

```

Comment:
Can you add a message to the assert here and above? There's not much distinguishing a stack trace from a user otherwise.